File: 0078-subsets/0078-subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        sol = []

        def backtrack(i):
            if i == len(nums):
                res.append(sol[:])
                return
            
            backtrack (i + 1)
            
            sol.append(nums[i])
            backtrack(i + 1)
            sol.pop()

        backtrack(0)
        return res

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {

        List<List<Integer>> res = new ArrayList<>();
        List<Integer> sol = new ArrayList<>();
        backtrack(nums, 0, res, sol);

        return res;
        
    }

    private void backtrack(int[] nums, int i, List<List<Integer>> res, List<Integer> sol){

        if(i == nums.length){
            res.add(new ArrayList<>(sol));
            return;
        }

        sol.add(nums[i]);
        backtrack(nums, i + 1, res, sol);
        sol.remove(sol.size() - 1);
        backtrack(nums, i + 1, res, sol);
        
    }
}